cpp-client/deephaven/client/include/public/deephaven/client/client.h

```diff
@@ -1751,4 +1762,576 @@ TableHandle TableHandle::ungroup(bool nullFill, Args &&... columnSpecs) const {
   };
   return ungroup(nullFill, std::move(groupByColumns));
 }
+
+/**
+ * A UpdateByOperation represents an operator for the Table update-by operation.
+ */
+class UpdateByOperation {
+public:
+  /*
+   * Default constructor. Creates a (useless) empty client object.
+   */
+  UpdateByOperation();
+  /**
+   * Constructor. Used internally.
+   */
+  explicit UpdateByOperation(std::shared_ptr<impl::UpdateByOperationImpl> impl);
```

Comment:
public and no `struct Private` trick?

Reply:
No, this object is more like `Client` or `TableHandle` in that I want it to act like a value type. The `Private` trick you are referring to is for objects that I want to force to be only createable as shared_ptrs, via a factory method. (In fact, the object owned by this object, namely `UpdateByOperationImpl` is itself such an class).

So this object has simple value type semantics, and is freely copyable, and so there might be multiple `UpdateByOperation` that point to the same `UpdateByOperationImpl` and that's fine.

Normally I might be a little more strict about this (for example, disallowing copies) but it's friendlier to R code and Cython code, and to some extend C++ code, if I just make these simple values.

I can possibly make this particular constructor private, and use `friend` but I don't like `friend` very much. And I thought, there really isn't much harm callers can do if they try to call this constructor. But let me see what I can do about hiding it.

Reply:
Well, I'm not sure what to do here. I'd like to have some concept like "package private" but that doesn't exist in C++. I might leave it alone for now pending further discussion.

Reply:
SG, no worries, no need for further action, explanation makes sense.